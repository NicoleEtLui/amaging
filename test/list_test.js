/* eslint-env jest */

import request from 'supertest'
import crypto from 'crypto'

import appFactory from './fixtures/app'
let app = null

const ACCESS_KEY = 'apiaccess'
const SECRET_KEY = '4ec2b79b81ee67e305b1eb4329ef2cd1'

beforeAll(done => { app = appFactory(done) })

/*
        LIST FILES IN A DIR
*/

describe('GET a file', () => {
  const createPolicy = (secret, data) => {
    data.expiration = new Date(new Date().getTime() + 1000 * 60)
    const policy = Buffer.from(JSON.stringify(data)).toString('base64')

    const sign = crypto.createHmac('sha1', secret)
    sign.update(policy)
    const token = sign.digest('hex').toLowerCase()

    return {
      policy,
      token
    }
  }

  const listTest = async (path, policyConditions = []) => {
    const policy = createPolicy(SECRET_KEY, {
      conditions: policyConditions
    })

    const res = await request(app)
      .get(`/test${path}`)
      .set('x-authentication', ACCESS_KEY)
      .set('x-authentication-token', policy.token)
      .set('x-authentication-policy', policy.policy)
      .expect(200)
    expect(res.body).toEqual(expect.any(Array))
    res.body.forEach(file => {
      expect(file).toMatchSnapshot({
        LastModified: expect.any(String)
      })
    })
  }

  test('Should return a 403 because no auth', async () => {
    await request(app)
      .get(`/test/`)
      .expect(403)
  })
  test('Should return a 400 when policy does not allow `list` action', async () => {
    const policy = createPolicy(SECRET_KEY, {
      conditions: [
        ['eq', 'action', 'foo']
      ]
    })

    await request(app)
      .get(`/test/foo/bar/`)
      .set('x-authentication', ACCESS_KEY)
      .set('x-authentication-token', policy.token)
      .set('x-authentication-policy', policy.policy)
      .expect(400, 'Invalid value for key: action')
  })
  test('Should return a 400 when policy does not allow `key`', async () => {
    const policy = createPolicy(SECRET_KEY, {
      conditions: [
        ['eq', 'key', 'foo']
      ]
    })

    await request(app)
      .get(`/test/foo/bar/`)
      .set('x-authentication', ACCESS_KEY)
      .set('x-authentication-token', policy.token)
      .set('x-authentication-policy', policy.policy)
      .expect(400, 'Invalid value for key: key')
  })

  test('Should list files in storage in root', async () => {
    await listTest('/')
  })
  test('Should list files in storage in sub folder', async () => {
    await listTest('/sub-folder/')
  })
  test('Should list files when policy allow `list` action', async () => {
    await listTest('/sub-folder/', [
      ['eq', 'action', 'list']
    ])
    await listTest('/sub-folder/', [
      ['eq', 'action', 'list', 'create']
    ])
  })
  test('Should list files when policy allow `key`', async () => {
    await listTest('/sub-folder/', [
      ['eq', 'key', 'sub-folder/']
    ])
  })

  test('Should return a 404 error because of an unexpected url (access a directory without end slash)', async () => {
    await request(app)
      .get('/test')
      .expect(404)
  })

  test('Should return a 404 error because of an unexpected url (access a directory without end slash)', async () => {
    await request(app)
      .get('/test/a')
      .expect(404)
  })
})
