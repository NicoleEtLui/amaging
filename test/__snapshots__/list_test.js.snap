// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET a file Should list files in storage in root 1`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "cache-eviction-delete.jpg",
  "isDirectory": false,
  "path": "/cache-eviction-delete.jpg",
}
`;

exports[`GET a file Should list files in storage in root 2`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "cache-eviction-update.jpg",
  "isDirectory": false,
  "path": "/cache-eviction-update.jpg",
}
`;

exports[`GET a file Should list files in storage in root 3`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "delete.jpg",
  "isDirectory": false,
  "path": "/delete.jpg",
}
`;

exports[`GET a file Should list files in storage in root 4`] = `
Object {
  "ContentLength": 13,
  "ContentType": "application/json",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "file.json",
  "isDirectory": false,
  "path": "/file.json",
}
`;

exports[`GET a file Should list files in storage in root 5`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "ice.jpg",
  "isDirectory": false,
  "path": "/ice.jpg",
}
`;

exports[`GET a file Should list files in storage in root 6`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "igloo.jpg",
  "isDirectory": false,
  "path": "/igloo.jpg",
}
`;

exports[`GET a file Should list files in storage in root 7`] = `
Object {
  "ContentLength": 10,
  "ContentType": "application/octet-stream",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "some-file-with-unknown-ext.ozo",
  "isDirectory": false,
  "path": "/some-file-with-unknown-ext.ozo",
}
`;

exports[`GET a file Should list files in storage in root 8`] = `
Object {
  "ContentLength": 0,
  "ContentType": "application/x-directory",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "sub-folder",
  "isDirectory": true,
  "path": "/sub-folder",
}
`;

exports[`GET a file Should list files in storage in sub folder 1`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "ice.jpg",
  "isDirectory": false,
  "path": "/sub-folder/ice.jpg",
}
`;

exports[`GET a file Should list files when policy allow \`key\` 1`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "ice.jpg",
  "isDirectory": false,
  "path": "/sub-folder/ice.jpg",
}
`;

exports[`GET a file Should list files when policy allow \`list\` action 1`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "ice.jpg",
  "isDirectory": false,
  "path": "/sub-folder/ice.jpg",
}
`;

exports[`GET a file Should list files when policy allow \`list\` action 2`] = `
Object {
  "ContentLength": 17252,
  "ContentType": "image/jpeg",
  "ETag": Any<String>,
  "LastModified": Any<String>,
  "basename": "ice.jpg",
  "isDirectory": false,
  "path": "/sub-folder/ice.jpg",
}
`;
